comms/EclMeta.cpp

```diff
@@ -0,0 +1,309 @@
+        }
+        return retval;
+    }
+
+    virtual HRESULT STDMETHODCALLTYPE endElement(
+        const wchar_t *pwchNamespaceUri,
+        int cchNamespaceUri,
+        const wchar_t *pwchLocalName,
+        int cchLocalName,
+        const wchar_t *pwchQName,
+        int cchQName)
+    {
+        Element &e = m_stack.top();
+        if (e.m_tag.compare(_T("Definition")) == 0) {
+            CEclDefinition *def = m_defStack.top();
+            def->UpdateAttrs(e);
+            if (m_currSource)
+                m_currSource->UpdateDef(def);
+            m_defStack.pop();
+        }
+        else if (e.m_tag.compare(_T("Field")) == 0) {
+            CEclDefinition *def = m_defStack.top();
+            def->AddField(e);
+        }
+        else if (e.m_tag.compare(_T("Import")) == 0) {
+            CEclImport *import = m_importStack.top();
+            import->UpdateAttrs(e);
+            if (m_currSource)
+                m_currSource->UpdateImport(import);
+            m_importStack.pop();
+        }
+
+        return __super::endElement(pwchNamespaceUri, cchNamespaceUri, pwchLocalName, cchLocalName, pwchQName, cchQName);
+    }
+};
+
+CEclFile *CEclMeta::GetSourceFromPath(const std::wstring & path)
+{
+    for (std::map<std::wstring, StlLinked<CEclMetaData>>::iterator itr = m_masterMeta.begin(); itr != m_masterMeta.end(); ++itr)
+    {
+        CEclMetaData *meta = itr->second;
+        if (meta != NULL && meta->HasPath())
+        {
+            std::_tstring pather = meta->GetPath().c_str();
+            if (boost::algorithm::equals(pather, path))
+            {
+                return (CEclFile *)meta;
+            }
+        }
+    }
+    return NULL;
+}
+
+void CEclMeta::Update(const std::wstring & xml)
+{
+    CComInitialize com;
+    CComPtr<ISAXXMLReader> reader;
+    HRESULT hr = reader.CoCreateInstance(CLSID_SAXXMLReader30);
+
+    CComPtr <CECLCCMetaHandler> handler = new CECLCCMetaHandler(this);
+    hr = reader->putContentHandler(handler);
+    if (hr == S_OK)
+    {
+        reader->parse(CComVariant(xml.c_str()));
+    }
+}
+
+void CEclMeta::LoadMetaData(const WPathVector *folders)
+{
+    for (WPathVector::const_iterator itr = folders->begin(); itr != folders->end(); ++itr)
+    {
+        PopulateMeta(itr->first,_T(""));
+    }
+}
+
+void CEclMeta::PopulateMeta(const boost::filesystem::wpath & fileOrDir, const std::wstring dottedPath, int level)
+{
+    if (clib::filesystem::exists(fileOrDir))
+    {
+        std::wstring newDottedPath;
+        
+        if (level)
+        {
+            if (dottedPath.empty())
+                newDottedPath = fileOrDir.stem().c_str();
+            else
+                newDottedPath = dottedPath + _T(".") + fileOrDir.stem().c_str();
+        }
+
+        if (HasValidExtension(pathToWString(fileOrDir)))
+        {
+            m_masterMeta[newDottedPath] = new CEclFile(fileOrDir);
+        }
+        else if (boost::filesystem::is_directory(fileOrDir))
```

Comment:
Might be better to test if its a Dir first?

Reply:
agreed